### PR TITLE
Extract Common Listing and Retrieval Functionality

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

Extracts common listing and get functionality from the various cloud ObjectStore. This ends up being more code, but avoids duplication and so I think is probably better, would appreciate any thoughts?

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/aws/client.rs [hunk 1]
```diff
@@ -302,89 +268,6 @@ impl S3Client {
         Ok(())
     }
 
-    /// Make an S3 List request <https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html>
-    async fn list_request(
```
Comment (tustvold) [inline object_store/src/aws/client.rs:306]: This is just moved

File: object_store/src/aws/client.rs [hunk 2]
```diff
@@ -169,40 +169,6 @@ impl S3Client {
         self.config.credentials.get_credential().await
     }
 
-    /// Make an S3 GET request <https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html>
-    pub async fn get_request(
```
Comment (tustvold) [inline object_store/src/aws/client.rs:173]: This is moved to be a trait implementation

File: object_store/src/aws/client.rs [hunk 3]
```diff
@@ -302,89 +268,6 @@ impl S3Client {
...
-
-        let mut response: ListResponse = quick_xml::de::from_reader(response.reader())
-            .context(InvalidListResponseSnafu)?;
-        let token = response.next_continuation_token.take();
-
-        Ok((response.try_into()?, token))
-    }
-
-    /// Perform a list operation automatically handling pagination
-    pub fn list_paginated(
```
Comment (tustvold) [inline object_store/src/aws/client.rs:363]: This is now implemented by ListClientExt

File: object_store/src/azure/client.rs [hunk 4]
```diff
@@ -300,14 +267,59 @@ impl AzureClient {
...
+impl ListClient for AzureClient {
     /// Make an Azure List request <https://docs.microsoft.com/en-us/rest/api/storageservices/list-blobs>
     async fn list_request(
         &self,
         prefix: Option<&str>,
         delimiter: bool,
         token: Option<&str>,
+        offset: Option<&str>,
     ) -> Result<(ListResult, Option<String>)> {
+        assert!(offset.is_none()); // Not yet supported
```
Comment (alamb) [inline object_store/src/azure/client.rs:321]: This should return an error rather than panic'ing I think

File: object_store/src/client/get.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+use crate::client::header::header_meta;
+use crate::path::Path;
+use crate::Result;
+use crate::{Error, GetOptions, GetResult, ObjectMeta};
+use async_trait::async_trait;
+use futures::{StreamExt, TryStreamExt};
+use reqwest::Response;
+
+#[async_trait]
+pub trait GetClient: Send + Sync + 'static {
```
Comment (alamb) [inline object_store/src/client/get.rs:27]: I think some this should have some comments / explination for what it is doing

File: object_store/src/azure/client.rs [hunk 6]
```diff
@@ -346,22 +358,6 @@ impl AzureClient {
 
         Ok((to_list_result(response, prefix)?, token))
     }
-
-    /// Perform a list operation automatically handling pagination
```
Comment (alamb) [inline object_store/src/azure/client.rs:350]: This is now implemented indirectly via `ListClientExt`

**Review by alamb — APPROVED** [review]
The structure looks (really) nice to me. 🏅  work @tustvold 

I think the new traits could do with some docs (maybe just pointing at the `ObjectStore::` trait to explain their semantics) but otherwise 🚀 I think

File: object_store/src/azure/client.rs (hunk 4 above)
Comment (tustvold) [reply object_store/src/azure/client.rs:321]: It is unreachable by the current code

**Pull request merged**